### ISSUE TITLE
Pushing to docker-hub is very slow, for existing images.

### ISSUE THREAD

**Issue by andersjanmyr** [docker/docker#13454]
Hi,

Pushing images is very slow, for example pushing an image with 25 already existing layers takes
me about 3 minutes.

It seems to me that there are two problems with the current implementation.
* No caching of the the existing checksums leads to a redirect on every HEAD request.
* No bulk-check is implemented

The client knows all the checksums it want to send and it should be possible to implement a bulk check endpoint that can take a list of checksums and reply with the ones that the server needs.

I don't know if this is the correct place to report this, and in that case I apologize and ask you to point me to a better place to discuss this.

**Comment by thaJeztah** [comment]
This might be a duplicate of https://github.com/docker/docker/issues/13250, but perhaps you can check if you think that's the case. The discussion on that issue may give you some pointers on what to check and try :)

**Comment by andersjanmyr** [comment]
@thaJeztah I don't believe this is a duplicate. My case is about pushing existing images which should need very little IO with a proper Cache. It is also about extending the v2 protocol with additional functionality for checking many images with a single request.

**Comment by thaJeztah** [comment]
Discussions around the V2 protocol may be better suited in the https://github.com/docker/distribution repository, but I'll ask one of the maintainers to have a look here briefly, perhaps this is something that's being worked on already;

ping @stevvooe this is about optimizing pushing of *existing* images

**Comment by andersjanmyr** [comment]
@thaJeztah I believe you are right, it should be moved to https://github.com/docker/distribution.

**Comment by andersjanmyr** [comment]
Transferred to the distribution repo, https://github.com/docker/distribution/issues/563

**Comment by stevvooe** [comment]
@andersjanmyr Actually, these may be bugs in docker, rather than distribution. Let's keep the conversation here until we've isolated the issue.

What behavior are you seeing that led you to these particular conclusions? Do you have data to isolate these specific items as the cause of the slowness?

What phase of the push are you seeing the slow push? I'd recommend reading the existing issues carefully to ensure that they are not the same. One is characterized by high CPU while the other is driven by a slow registry backend.

In response to the individual items:

1. What do you mean by "no caching of the existing checksums"? Is this in the registry client or the registry itself? There is some caching going on here on the server side but it's unclear what you think is the actual problem.
2. The value of bulk check is quite dubious without measuring performance. There are a host of other optimizations that could be done before adding a bulk check. Indeed, there are some very cool algorithms for syncing sets of content-addressable content, but, right now, the single HEAD checks are not the bottleneck.

While these may seem to be simple observations to make, implementing these optimizations will not always lead to a noticeable performance improvement without careful analysis. It will also serve to make the code more complex without tangible benefit. For instance, issuing the HEAD requests in parallel may be much faster if done correctly.

If this is simply an enhancement request, let's try to identify which enhancements will have an effect and prioritize them.

@thaJeztah In the future, let's make sure to fully triage the issue before dispatching to another project.

**Comment by andersjanmyr** [comment]
@stevvooe Yes, you are right that there may be a number of better solutions than what I have suggested. I agree that it is important to keep the code as simple as possible, but one of the things that I love about Docker is the speed, and it seems to me that the networking speed can be improved. Making 25 requests when one will suffice doesn't seem right.

When I first started looking at this I thought that Docker uses a similar process to Git.
https://git-scm.com/book/en/v2/Git-Internals-Transfer-Protocols#The-Smart-Protocol

This is how I came (jumped :) ) to my conclusions:

First of all the command line output, showing that a push of an existing image takes 60 seconds
docker push andersjanmyr/pinga
The push refers to a repository [andersjanmyr/pinga] (len: 1)
2fa5b9baa10c: Image already exists
20 layers truncated...
41b730702607: Image already exists
3cb35ae859e7: Image already exists
Digest: sha256:c6fc6082e22c95000244f4dc4cd3158c2fd56f9618a8c3e58b7d8e9e92ab644d
[5413:andersjanmyr:pinga 3c25dfe@master *=60s]

Then I have the debug log from the docker daemon, it shows that each HEAD request takes between 1 and 5 seconds.
https://gist.github.com/andersjanmyr/ea1f46ca51d1d12f35dd

Truncated excerpt from above log
time="2015-05-27T04:55:36Z" level=debug msg="[registry] Calling \"HEAD\" https://registry-1.docker.io/v2/andersjanmyr/pinga/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
time="2015-05-27T04:55:41Z" level=debug msg="[registry] Calling \"HEAD\"          https://
time="2015-05-27T04:55:42Z" level=debug msg="[registry] Calling \"HEAD\"          https://
time="2015-05-27T04:55:44Z" level=debug msg="[registry] Calling \"HEAD\"
time="2015-05-27T04:55:47Z" level=debug msg="[registry] Calling \"HEAD\"

Then I also started reading the code, and in blobserver.go I cannot find any caching at all
https://github.com/docker/distribution/blob/master/registry/storage/blobserver.go#L41
And, depending on what type of storage is used every call may also result in a redirect.

**Comment by dustinlacewell** [comment]
Also feeling the pain from this when trying to iterate with an image that only needs to push the last of about 30 layers on 1.8.1

**Comment by chekunkov** [comment]
To blow away the dust from this topic - wasn't the issue fixed by #18353? It should be part of `1.10.0` release https://github.com/docker/docker/milestones/1.10.0

**Comment by stevvooe** [comment]
@chekunkov Yes, that should have provided better layer reuse, in addition to other fixes that provided better metadata management. The main issue here is layer digest stabilization.

**Comment by LK4D4** [comment]
@stevvooe @aaronlehmann @tonistiigi do you think we can close this one?

**Comment by stevvooe** [comment]
@LK4D4 That should help, somewhat.

I am apt to close this issue, however, as the closure condition is slightly ambiguous. It has some suggestions of what to do without framing the problem. _When is pushing images to the hub no longer slow?_ If there are specific problems that are contributed to performance issues, we should file bugs for those and address those specifically.

The other problem I see in the framing of this issue is that the suggested fixes are unlikely to address the primary bottleneck. We should instrument the push/pull process and identify the actual issue. (To tell you the truth, I think there may be a UX bug, as well).

I'll close this, for now. If the original submitter would like some guidance on making metrics driven improvements, I think we can work together to find areas where gains can be made.
